File: back-end/src/main/java/com/action/controller/NewsUserController.java
package com.action.controller;

import com.action.common.Code;
import com.action.common.Result;
import com.action.entity.NewsUser;
import com.action.exception.BusinessException;
import com.action.service.NewsUserService;
import com.action.util.JwtUtil;
import com.action.util.MD5Util;
import com.action.util.ThreadLocalUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/users")
public class NewsUserController {


    @Autowired
    private NewsUserService userService;

    /**
     * 检查用户名是否已存在
     * @param username
     * @return
     */
    @GetMapping("/check/{username}")
    public Result checkUserName(@PathVariable String username) {
        NewsUser newsUser = userService.findByUsername(username);
        if (newsUser != null) {
            return Result.error(Code.USERNAME_USED, "用户名已存在");
        }
        return Result.ok(null);
    }

    /**
     * 根据 uid 获取用户信息
     * @return
     */
    @GetMapping("/info")
    public Result getUserInfo() {
        int userId = ThreadLocalUtil.getUserId();
        NewsUser user = userService.findByUid(userId);
        if (user != null) {
            Map<Object, Object> data = new HashMap<>();
            user.setUserPwd(null);
            data.put("user", user);
            return Result.ok(data);
        }
        return Result.error(Code.USER_NOT_EXIST, "用户不存在");
    }

    /**
     * 用户登录
     * @param loginUser
     * @return
     */
    @PostMapping("/login")
    public Result login(@RequestBody NewsUser loginUser) {
        if (loginUser == null) {
            throw new BusinessException("用户登录参数不能为空", Code.REQUEST_ERROR);
        }

        NewsUser user = userService.findByUsername(loginUser.getUsername());
        Result result = Result.error(Code.USERNAME_OR_PASSWORD_ERROR, "用户名或密码错误");
        // 验证用户名密码
        if (user != null && MD5Util.encrypt(loginUser.getUserPwd()).equalsIgnoreCase(user.getUserPwd())) {
            Map<Object, Object> data = new HashMap<>();
            data.put("token", JwtUtil.createToken(user.getUid()));
            result = Result.ok(data);
        }

        return result;
    }


    /**
     * 用户注册
     * @param registerUser
     * @return
     */
    @PostMapping("/register")
    public Result register(@RequestBody NewsUser registerUser) {
        if (registerUser == null) {
            throw new BusinessException("用户注册参数不能为空", Code.REQUEST_ERROR);
        }

        userService.register(registerUser);
        return Result.ok(null);
    }
}

File: back-end-gin/controller/news_user_controller.go
package controller

import (
	"github.com/action/micro-headlines/common"
	"github.com/action/micro-headlines/models"
	"github.com/action/micro-headlines/service"
	"github.com/action/micro-headlines/utils"
	"github.com/action/micro-headlines/vo"
	"github.com/gin-gonic/gin"
)

type NewsUserController struct {
	userService *service.NewsUserService
}

func NewNewsUserController(userService *service.NewsUserService) *NewsUserController {
	return &NewsUserController{userService: userService}
}

// CheckUserName 检查用户名是否已存在
func (ctrl *NewsUserController) CheckUserName(c *gin.Context) {
	username := c.Param("username")

	user, err := ctrl.userService.FindByUsername(username)
	if err != nil {
		common.Error(c, common.SYSTEM_UNKNOWN_ERROR, err.Error())
		return
	}

	if user != nil {
		common.Error(c, common.USERNAME_USED, "用户名已存在")
		return
	}

	common.Ok(c, nil)
}

// GetUserInfo 获取用户信息
func (ctrl *NewsUserController) GetUserInfo(c *gin.Context) {
	userId, exists := utils.GetUserId(c)
	if !exists {
		common.Unauthorized(c)
		return
	}

	user, err := ctrl.userService.FindByUid(userId)
	if err != nil {
		common.Error(c, common.SYSTEM_UNKNOWN_ERROR, err.Error())
		return
	}

	if user == nil {
		common.Error(c, common.USER_NOT_EXIST, "用户不存在")
		return
	}

	// 隐藏密码
	user.UserPwd = ""

	data := map[string]interface{}{
		"user": user,
	}

	common.Ok(c, data)
}

// Login 用户登录
func (ctrl *NewsUserController) Login(c *gin.Context) {
	var req vo.UserRequestVo

	if err := c.ShouldBindJSON(&req); err != nil {
		common.Error(c, common.REQUEST_ERROR, "用户登录参数不能为空: "+err.Error())
		return
	}

	user, err := ctrl.userService.FindByUsername(req.Username)
	if err != nil {
		common.Error(c, common.SYSTEM_UNKNOWN_ERROR, err.Error())
		return
	}

	// 验证用户名密码
	if user == nil || utils.Encrypt(req.UserPwd) != user.UserPwd {
		common.Error(c, common.USERNAME_OR_PASSWORD_ERROR, "用户名或密码错误")
		return
	}

	// 生成 token
	token, err := utils.CreateToken(user.Uid)
	if err != nil {
		common.Error(c, common.SYSTEM_UNKNOWN_ERROR, "生成 token 失败")
		return
	}

	data := map[string]interface{}{
		"token": token,
	}

	common.Ok(c, data)
}

// Register 用户注册
func (ctrl *NewsUserController) Register(c *gin.Context) {
	var req vo.UserRequestVo

	if err := c.ShouldBindJSON(&req); err != nil {
		common.Error(c, common.REQUEST_ERROR, "用户注册参数不能为空: "+err.Error())
		return
	}

	// 转换为数据库模型
	registerUser := &models.NewsUser{
		Username: req.Username,
		UserPwd:  req.UserPwd,
		NickName: req.NickName,
	}

	err := ctrl.userService.Register(registerUser)
	if err != nil {
		common.Error(c, common.REGISTER_ERROR, err.Error())
		return
	}

	common.Ok(c, nil)
}
